lib/skate/detours/route_segments.ex

```diff
@@ -21,49 +21,73 @@ defmodule Skate.Detours.RouteSegments do
+      ...>     Location.new(0.0, 1.0),
+      ...>     Location.new(0.0, 2.0),
+      ...>     Location.new(0.0, 3.0)
+      ...>   ],
+      ...>   Location.new(0.0, 1.3),
+      ...>   Location.new(0.0, 2.6)
       ...> )
       {:ok,
         %Skate.Detours.RouteSegments.Result{
-          before_detour: [%Util.Location{latitude: 0, longitude: 0}, %Util.Location{latitude: 0, longitude: 1}],
-          detour: [%Util.Location{latitude: 0, longitude: 1}, %Util.Location{latitude: 0, longitude: 2}],
-          after_detour: [%Util.Location{latitude: 0, longitude: 2}, %Util.Location{latitude: 0, longitude: 3}]
+          before_detour: [
+            %Util.Location{latitude: 0.0, longitude: 0.0},
+            %Util.Location{latitude: 0.0, longitude: 1.0},
+            %Util.Location{latitude: 0.0, longitude: 1.3}
+          ],
+          detour: [
+            %Util.Location{latitude: 0.0, longitude: 1.3},
+            %Util.Location{latitude: 0.0, longitude: 2.0},
+            %Util.Location{latitude: 0.0, longitude: 2.6}
+          ],
+          after_detour: [
+            %Util.Location{latitude: 0.0, longitude: 2.6},
+            %Util.Location{latitude: 0.0, longitude: 3.0}
+          ]
         }}
   """
   @spec route_segments(
           nonempty_list(Util.Location.From.t()),
           Util.Location.From.t(),
           Util.Location.From.t()
         ) :: {:ok, Result.t()} | :error
-  def route_segments(
-        [],
-        _connection_start,
-        _connection_end
-      ),
-      do: :error
+  def route_segments([], _connection_start, _connection_end), do: :error
+  def route_segments([_], _connection_start, _connection_end), do: :error
 
   def route_segments(
         shape,
         connection_start,
         connection_end
       ) do
-    {_, start_index} =
-      shape
-      |> Enum.with_index()
-      |> Enum.min_by(fn {point, _index} -> Util.Location.distance(point, connection_start) end)
+    {nearest_start_point, start_index} =
+      nearest_point_to_shape(shape, connection_start)
 
-    {_, end_index} =
-      shape
-      |> Enum.with_index()
-      |> Enum.min_by(fn {point, _index} -> Util.Location.distance(point, connection_end) end)
+    {nearest_end_point, end_index} =
+      nearest_point_to_shape(shape, connection_end)
 
     {:ok,
      %__MODULE__.Result{
-       before_detour: Enum.slice(shape, 0..start_index),
-       detour: Enum.slice(shape, start_index..end_index),
-       after_detour: Enum.slice(shape, end_index..-1)
+       before_detour: Enum.slice(shape, 0..start_index) ++ [nearest_start_point],
+       detour:
+         [nearest_start_point] ++
+           Enum.slice(shape, (start_index + 1)..end_index) ++ [nearest_end_point],
+       after_detour: [nearest_end_point] ++ Enum.slice(shape, (end_index + 1)..-1)
      }}
   end
+
+  defp nearest_point_to_shape(shape, point) do
+    {%{nearest_point: nearest_point}, index} =
+      shape
+      |> Enum.zip(Enum.drop(shape, 1))
```

Comment:
question(non-blocking): is there a reason to prefer `zip` over [`chunk_every/4`](https://hexdocs.pm/elixir/main/Enum.html#chunk_every/2)?
```suggestion
      |> Enum.chunk_every(2, 1, :discard)
```

Reply:
I do see that `nearest_point_to_segment` takes a `Tuple` and not a `List` so that's one change, but I'm not sure it really matters?

Reply:
Huh - I didn't realize that `chunk_every` behaved that way. 

I imagine that I could do `chunk_every` and then transform the resulting array into a tuple before passing it into `nearest_point_to_segment`

Reply:
[Use `chunk_every`](https://github.com/mbta/skate/pull/2506/commits/2c97c4513e23ed10422ff27e5120d44ba7923ef3)
[Convert from array to tuple](https://github.com/mbta/skate/pull/2506/commits/323dc824b1bb319ea78a8dc3965847858789e24d)